Fixes for issue #90

maeriens opened the pull request:
Here I fix what I mentioned in #90, and also rebased from #89 so everything can go in one single PR.

maxhudson commented on src/index.js:
```diff
@@ -168,38 +170,41 @@ class StickyTable extends React.Component {
     //HINT avoid a render unless there's actually a change
     if (JSON.stringify(stickyInsets) !== JSON.stringify(this.state.stickyInsets)) {
-      this.setState({stickyInsets});
+      this.setState({ stickyInsets });
```
lots of opinionated styles mixed in with a functional pull request - I'm personally not a fan of the spacing style changes made in this PR, but do appreciate the functional code

maeriens replied:
I can revert them if you want - I think I have a global styling that somehow overwrote the `.eslint` or I used the format on save from vscode; not sure

maxhudson commented on src/index.js:
```diff
@@ -76,6 +76,8 @@ const Wrapper = styled('div').attrs({
     var insets = props.stickyInsets;
     var styles = '';
     var i;
+    console.log(props)
```
remove console logs

maxhudson commented on src/index.js:
```diff
@@ -24,17 +24,17 @@ const Cell = styled('div').attrs({
 Row.displayName = 'Row';
 
-const Wrapper = styled('div').attrs({
+const Wrapper = styled('div').attrs(() => ({
```
what's the reason for this change?

maeriens replied:
It was deprecated on `styled-components@4` but was backwards compatible - and dropped in v5 - it's point 7 in their FAQ here https://styled-components.com/docs/faqs#what-do-i-need-to-do-to-migrate-to-v4

maxhudson commented on src/index.js:
```diff
@@ -168,38 +170,41 @@ class StickyTable extends React.Component {
-      var insets = [0];
-      var count = props[countPropKey];
-      var netInset = 0;
+      var insets = [];
```
seems odd to have `insets[0]` be `undefined` for count 1, and `insets[0]` be 0 for count 2+

maxhudson replied:
implying this should be `var insets = [0]`

maeriens replied:
The first solution I had was to declare `var insets = props[countPropKey] > 1 ? [0] : [];` because the issue is that it sets to 0 all sides when any of the other stickyes is higher than 1

maxhudson replied:
I think my first comment in this thread is still relevant or the answer is unclear to me. I still don't understand why it should ever be empty. 

I actually wrote the code so that 0 would never be necessary here, so guessing insets should never have 0 in it, unless a cell has width 0. The first cell on a given side will always have an inset of 0 regardless of the status of the other sticky rows/columns

maxhudson replied:
seems like 0 should always be in insets, or never be in it

maxhudson commented on src/index.js:
```diff
@@ -168,38 +170,41 @@ class StickyTable extends React.Component {
-        }
+          if (node) {
+            var boundingRect = node.getBoundingClientRect();
+            netInset += boundingRect[sizeKey];
```
new line above 186

maxhudson commented on src/index.js:
```diff
@@ -168,38 +170,41 @@ class StickyTable extends React.Component {
-          var boundingRect = node.getBoundingClientRect();
+        // HINT we only want this loop for the second sticky and up
+        for (s = 1; s < count; s++) {
+          var node = stickyKey === 'header' || stickyKey === 'leftColumn' ? cellNodes[s - 1] : cellNodes[cellNodes.length - s];
```
this seems like the one significant change here - definitely should be doing something with s, and not sure how I missed that, but I'm pretty sure this isn't quite right. need to be doing something with rows and then cells both based on s (this is only accounting for more than 2 columns, but not more than 2 footers/headers)

maeriens replied:
Hmm. You're right. This fixes columns but not the others.
Maybe getting the `rowNodes` from the table too? That way, if the `stickyKey` is `header` or `footer`, it should check against the cells (probably `rowNodes[s - 1] : rowNodes[rowNodes.length - s];?)

maeriens commented:
@maxhudson do you want me to close this one and send again the change with only the storybook addons, and then in here continue the other changes?

maxhudson commented:
@maeriens I think it's fine to leave it in one place for now. Will give fixing the row/cell node indices a go tomorrow

maeriens commented:
Removed the spaces where they were unwanted - and added a patch that worked for all tests I ran (one sticky in one, two or three places, up to 4 stickies in one, two or three directions).
We can improve over this now - I didn't have time to give it a bit more of thought to come up with a stronger solution today

maxhudson commented on src/index.js:
```diff
@@ -168,21 +169,40 @@ class StickyTable extends React.Component {
+              node = rowNodes[s - 1]
+              break;
+            case 'footer':
+              node = rowNodes[rowNodes.length - s]
```
when the node is a row, we should be using childNodes[0] so we're always getting size of a cell - in case there's a browser bug with table row height

maxhudson commented:
I'm going to merge this as it's close enough and I'll play around with it/clean up on my end before publishing.

maxhudson commented:
Thank you very much for the help @maeriens  - let me know if I can buy you a coffee!

maxhudson commented:
@maeriens I couldn't get your playground to work - not sure if it's something on my end or just a misconfiguration. Assuming expected behavior is resizable column handles?

maeriens commented:
@maxhudson sorry I didn't answer before - no problemo over the coffee, I think this is my first real contribution to something useful. As for the playgorund, I just run `yarn run storybook` and it displays correctly the knobs in that story only (not in the rest). Here is a gif of it.
![image](https://user-images.githubusercontent.com/5074186/81721567-6de45400-9456-11ea-86b3-bf9d408b3ef9.png)

maxhudson commented:
Weird - not sure how to get that to show up. Doesn't seem to be working here: https://henrybuilt.github.io/react-sticky-table/?path=/story/basic--playground

maxhudson commented:
@maeriens out of curiosity - are you using the package for something? Would be great to have some example links!

maeriens commented:
Ah - I think the tab is not open by default - press `D` in the keyboard to bring it up. It an be configured, I forgot

maxhudson commented:
ah cool - thanks
